Add state change details to event reference (#2772)
sensu/sensu-docs: content/sensu-go/6.1/observability-pipeline/observe-events/events.md

## Use event data

Observability data in events is a powerful tool for automating monitoring workflows.

> For example, you can use the [`state` attribute][36] to provide handlers with more information about changes in check status and [reduce alert fatigue][23] by filtering events based on the event `occurrences` attribute.

Reviewer: ```suggestion
For example, the [`state` attribute][36] provides handlers with a high-level description of check status. Filtering events based on this attribute can help [reduce alert fatigue][23].
```

### State attribute

The `state` event attribute adds meaning to the check status:

- `passing` means the check status is `0` (OK).

> - `failing` means the check status changed from 0 to non-zero (WARNING or CRITICAL).

Reviewer: ```suggestion
- `failing` means the check status is non-zero (WARNING or CRITICAL).
```

I believe that events for a entity/check pair which has never reported an OK status are still considered to be a `failing` state

> - `flapping` means a rapid change in check result status (determined based on the [low and high flap thresholds][37] set in the check attributes).

Reviewer: ```suggestion
- `flapping` means a rapid change in check result status (determined based on per-check [low and high flap thresholds][37] attributes).
```

Reviewer 2: I'm not sure that rapid is the right word here. Rather, flapping conveys an unsteady state where the check is not settling on success or failure, in aggregate, over the last 21 executions.

> Flapping indicates problems with an entity, provided your low and high flap threshold settings are properly configured.

Reviewer: ```suggestion
Flapping typically indicates intermittent problems with an entity, provided your low and high flap threshold settings are properly configured.
Although some teams choose to filter out flapping events to reduce unactionable alerts, we suggest sending flapping events to a designated handler for later review.
```

> Tuning your flap threshold configuration based on Sensu's [flap detection algorithm][39] allows you to take action based on `flapping` state check results.

Reviewer: ```suggestion
If you repeatedly observe events in flapping state, Sensu's per-check flap threshold configuration allows you to adjust the sensitivity of the [flap detection algorithm][39].
```

#### Flap detection algorithm

Sensu uses the same flap detection algorithm as [Nagios][38].
Every time you run a check, Sensu records whether the `status` value changed since the previous check.

> Sensu stores the last 21 `status` values and uses them to calculate an entity's percent state change.

Reviewer: ```suggestion
Sensu stores the last 21 `status` values and uses them to calculate the percent state change for the entity/check pair.
```

Reviewer: The algorithm is applied to entity/check pairs. As written, some of this seems to be describing percent state change across all check results for an entity.

Then, Sensu's algorithm applies a weight to these status changes: more recent changes have more value than older changes.

After calculating the weighted total percent state change, Sensu compares it with the [low and high flap thresholds][37] set in the check attributes.

> - If the entity was **not** already flapping and the weighted total percent state change is greater than or equal to the `high_flap_threshold` setting, the entity has started flapping.

Author: ```suggestion
- If the entity was **not** already flapping and the weighted total percent state change for the entity/check pair is greater than or equal to the `high_flap_threshold` setting, the entity has started flapping.
```

> - If the entity **was** already flapping and the weighted total percent state change is less than the `low_flap_threshold` setting, the entity has stopped flapping.

Author: ```suggestion
- If the entity **was** already flapping and the weighted total percent state change for the entity/check pair is less than the `low_flap_threshold` setting, the entity has stopped flapping.
```

## Use event data

Observability data in events is a powerful tool for automating monitoring workflows.
For example, the [`state` attribute][36] provides handlers with a high-level description of check status. Filtering events based on this attribute can help [reduce alert fatigue][23].

### State attribute

The `state` event attribute adds meaning to the check status:

- `passing` means the check status is `0` (OK).
- `failing` means the check status is non-zero (WARNING or CRITICAL).

> - `flapping` means a rapid change in check result status (determined based on per-check [low and high flap thresholds][37] attributes).

Author: ```suggestion
- `flapping` indicates an unsteady state in which the check result status (determined based on per-check [low and high flap thresholds][37] attributes) is not settling on `passing` or `failing` according to the [flap detection algorithm][39].
```

## Use event data

Observability data in events is a powerful tool for automating monitoring workflows.
For example, you can use the [`state` attribute][36] to provide handlers with more information about changes in check status and [reduce alert fatigue][23] by filtering events based on the event `occurrences` attribute.

### State attribute

The `state` event attribute adds meaning to the check status:

- `passing` means the check `status` is 0 (OK).
- `failing` (status other than 0) means the check `status` changed from 0 to non-zero (WARNING or CRITICAL).
- `flapping` indicates a rapid change in check result status (determined by the [low and high flap thresholds][37] set in the check attributes).

#### Flap detection algorithm

Flapping can indicate network or entity problems, but it can also indicate improper [flap threshold][37] configuration.

> Learning more about Sensu's flap detection algorithm will allow you to properly tune your flap threshold configuration and be confident in `flapping` state check results.

Author: ```suggestion
Learning more about Sensu's flap detection algorithm will allow you to properly tune your flap threshold configuration and take appropriate action based on `flapping` state check results.
```

Sensu uses the same flap detection algorithm as [Nagios][38].
Every time you run a check, Sensu records whether the `status` value changed since the previous check.
The flap detection algorithm uses the number of `status` value changes for the last 21 checks to determine an entity's overall percent state change.
However, the algorithm weights these status changes differently: more recent changes have 50% more weight than older changes.

> Read the [Nagios Detection and Handling of State Flapping][38] guide for detailed flap detection logic calculation.

Author: ```suggestion
Read the [Nagios Detection and Handling of State Flapping][38] guide for detailed information about the flap detection calculation.
```

After calculating the weighted percent state change, Sensu compares this value with the [low and high flap thresholds][37] set in the check attributes:

- If the entity was **not** already flapping and the weighted percent state change is greater than or equal to the `high_flap_threshold` setting, Sensu determines that the entity has started flapping.
- If the entity **was** already flapping and the weighted percent state change is less than the `low_flap_threshold` setting, Sensu determines that the entity has stopped flapping.
- If neither, Sensu determines no change: either the entity has not started flapping or is still flapping.

If an entity has started flapping, Sensu will:

> - Log a message indicating that the entity is flapping.

Author: ```suggestion
- Log a message that the entity is flapping.
```

- Add a non-persistent comment to the entity indicating that it is flapping.
- Send a "flapping start" notification for the entity to appropriate contacts.
- Silence other notifications for the entity (if set as an event filter).

If an entity has stopped flapping, Sensu will:

> - Log a message indicating that the entity has stopped flapping.

Author: ```suggestion
- Log a message that the entity has stopped flapping.
```
